examples/coappearances.cpp

```diff
@@ -0,0 +1,327 @@
+            },
+            [strings]( co::Description desciption ) {
+                std::cout << ": Description: " << strings + desciption.ref;
+            },
+            [strings, vertices]( co::UnaryRelation relation ) {
+                std::cout << ": Relation(kind=" << strings + relation.kind_ref
+                          << ", to=" << strings + vertices[ relation.to_ref ].name_ref << ")";
+            },
+            [strings, vertices]( co::BinaryRelation relation ) {
+                std::cout << ": Relation(kind=" << strings + relation.kind_ref
+                          << ", to=" << strings + vertices[ relation.to_a_ref ].name_ref
+                          << ", to=" << strings + vertices[ relation.to_b_ref ].name_ref << ")";
+            } );
+        graph.vertices_data( ).for_each( vertex_ref, visitor );
+        std::cout << std::endl;
+    }
+    std::cout << std::endl;
+
+    std::cout << "Edges: " << std::endl;
+    auto edges = graph.edges( );
+    // Skip the last edge since it is a sentinel
+    for ( uint32_t edge_ref = 0; edge_ref + 1 < edges.size( ); ++edge_ref )
+    {
+        auto edge = edges[ edge_ref ];
+        std::cout << strings + vertices[ edge.a_ref ].name_ref << " meets "
+                  << strings + vertices[ edge.b_ref ].name_ref << " " << edge.count
+                  << " time(s) in chapters ";
+        // The end of the chapters assigned to this edge is the first chapter from the next edge.
+        // This is a typical trick when storing ranges. That's why a sentinel was added to edges.
+        uint32_t next_edge_ref = edge_ref + 1;
+        uint32_t chapters_begin = edge.first_chapter_ref;
+        uint32_t chapters_size = edges[ next_edge_ref ].first_chapter_ref - chapters_begin;
+        auto chapters = graph.chapters( ).slice( chapters_begin, chapters_size );
+        for ( size_t chapter_ref = 0; chapter_ref < chapters.size( ); ++chapter_ref )
+        {
+            auto chapter = graph.chapters( )[ chapter_ref ];
+            std::cout << chapter.major.as< uint32_t >( ) << "." << chapter.minor.as< uint32_t >( )
+                      << ( chapter_ref + 1 < chapters.size( ) ? ", " : "." );
+        }
+        std::cout << std::endl;
+    }
+}
+
+static const char* USAGE = 1 + R"_(
+Usage:
+  coappearances convert <input.json> <output_folder>
+  coappearances read <output_folder>)_";
+
+int
+main( int argc, char const* argv[] )
+{
+    if ( argc < 3 )
+    {
+        std::cerr << USAGE << std::endl;
+        return 1;
+    }
+
+    std::string verb( argv[ 1 ] );
+    try
+    {
+        if ( verb == "convert" && argc == 4 )
+        {
+            convert( argv[ 2 ], argv[ 3 ] );
+        }
+        else if ( verb == "read" && argc == 3 )
+        {
+            read( argv[ 2 ] );
+        }
+        else
+        {
+            std::cerr << USAGE << std::endl;
+            return 1;
+        }
+    }
+    catch ( std::runtime_error err )
```

Comment:
When catching an exception, catch by `const&`. https://stackoverflow.com/questions/2145147/why-catch-an-exception-as-reference-to-const

Reply:
Good catch.